<a href="https://colab.research.google.com/github/AxelYoel/FlyRank-AI-Internship---Axel-Yoel-Chandra/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/AxelYoel/FlyRank-AI-Internship---Axel-Yoel-Chandra/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

I'm choosing **Lane 1: Ranking Signal Analysis**. Before I try to build a
ranked action queue (Lane 2) or find content archetypes (Lane 3), I want to
first understand which observable signals — position, freshness, word count,
engagement, CTR — actually correlate with a page's visibility, clicks, or
movement, and which ones don't. This feels like the right starting point
because the other lanes implicitly assume you already know which signals
matter; Lane 1 is where I'd find that out rather than guess. It's also a
natural fit for the starter dataset and the warehouse's daily fact table,
since both are strongest on exactly these kinds of observed search and
engagement metrics

In [15]:
!git clone https://github.com/AxelYoel/FlyRank-AI-Internship---Axel-Yoel-Chandra.git
%cd FlyRank-AI-Internship---Axel-Yoel-Chandra

import pandas as pd
data = pd.read_csv("data/raw/content_refresh_anonymized.csv")
data

Cloning into 'FlyRank-AI-Internship---Axel-Yoel-Chandra'...
remote: Enumerating objects: 124, done.
remote: Counting objects: 100% (124/124), done.
remote: Compressing objects: 100% (81/81), done.
remote: Total 124 (delta 36), reused 92 (delta 27), pack-reused 0 (from 0)
Receiving objects: 100% (124/124), 1.99 MiB | 10.99 MiB/s, done.
Resolving deltas: 100% (36/36), done.
/content/FlyRank-AI-Internship---Axel-Yoel-Chandra/FlyRank-AI-Internship---Axel-Yoel-Chandra/FlyRank-AI-Internship---Axel-Yoel-Chandra/FlyRank-AI-Internship---Axel-Yoel-Chandra


,content_id,client_id,search_volume,competition,competition_level,cpc,content_type,main_intent,word_count,char_count,...,char_count_tier,ctr,avg_position,engagement_rate,scroll_rate,ai_traffic_pct,impression_tier,position_tier,trend_direction,trend_pct
0,content_304f48230142,client_f369cb89fc,10.0,0.67,HIGH,2.05,keyword article,transactional,3221.0,20457.0,...,15000-25000,0.76,10.6,5.88,4.55,0.00,good,striking,down,-41.4
1,content_a1fb4e703a9e,client_4e07408562,90.0,0.01,LOW,0.05,keyword article,informational,2481.0,15562.0,...,15000-25000,0.05,20.3,0.00,10.00,0.00,good,page_3_5,down,-57.7
2,content_9aa793d4d895,client_7f2253d7e2,0.0,0.00,LOW,0.00,keyword article,informational,3515.0,23643.0,...,15000-25000,0.09,36.5,0.00,28.57,0.00,good,page_3_5,down,-60.9
3,content_331d6c4de07b,client_19581e27de,10.0,0.00,LOW,0.00,keyword article,commercial,NaN,NaN,...,NaN,0.49,6.2,1.28,3.45,0.00,good,page_1,stable,-13.8
4,content_d99b7a2d90ca,client_3fdba35f04,0.0,0.00,LOW,0.00,keyword article,informational,2803.0,17469.0,...,15000-25000,0.13,44.0,0.00,24.29,0.00,good,page_3_5,down,-34.7
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
29995,content_c322796023c8,client_e29c9c180c,10.0,0.05,LOW,0.00,keyword article,transactional,1386.0,9084.0,...,8000-15000,0.00,0.0,0.00,0.00,0.00,low,top_3,new,NaN
29996,content_526572edb3fa,client_7f2253d7e2,0.0,0.00,LOW,0.00,keyword article,informational,2654.0,17056.0,...,15000-25000,0.39,6.6,0.00,66.67,0.00,moderate,page_1,down,-75.1
29997,content_38112bdd0c6e,client_349c41201b,10.0,1.00,HIGH,0.00,keyword article,transactional,2857.0,18725.0,...,15000-25000,0.19,4.1,0.00,0.00,0.00,good,page_1,down,-66.2
29998,content_ab26273a7e7a,client_19581e27de,10.0,0.00,LOW,0.00,keyword article,transactional,NaN,NaN,...,NaN,0.22,6.0,1.73,4.06,0.00,excellent,page_1,down,-27.9


In [10]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 44 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   content_id              30000 non-null  object 
 1   client_id               30000 non-null  object 
 2   search_volume           27532 non-null  float64
 3   competition             27532 non-null  float64
 4   competition_level       27390 non-null  object 
 5   cpc                     27532 non-null  float64
 6   content_type            30000 non-null  object 
 7   main_intent             27626 non-null  object 
 8   word_count              22301 non-null  float64
 9   char_count              22301 non-null  float64
 10  provider_used           8562 non-null   object 
 11  model_used              24267 non-null  object 
 12  impressions_90d         30000 non-null  int64  
 13  clicks_90d              30000 non-null  int64  
 14  pageviews_90d           30000 non-null

## 2. The question: decision, action, cost of a wrong call

My question: which observable content and search signals (position,
freshness, word count, engagement, CTR, etc.) are actually associated with
a page's visibility, clicks, or engagement — and how strong is each
association?

**Decision this improves:** right now, a content or SEO team has to guess
or rely on general SEO folklore when deciding what to watch or prioritize
across a large content inventory. This project replaces that guesswork
with evidence about which signals actually move with performance in
FlyRank's own data.

**Who acts on it:** a content strategist or SEO analyst, who uses the
signal report to decide what to monitor and where to focus review effort
— for example, whether freshness matters more than word count, or whether
position decay tends to show up before click declines.

**Cost of a wrong call:** since this is a signal-discovery project rather
than a direct action queue, the main risk is misdirected attention rather
than a bad one-off decision. If I claim a signal is meaningfully
associated with performance when it isn't (or miss one that actually
matters), teams could spend time optimizing or tracking the wrong thing —
for instance, obsessing over word count while a stronger signal like
position or freshness goes unnoticed. Because this is observational, not
causal, the bigger risk is someone treating a correlation as a guaranteed
lever to pull, rather than a pattern worth investigating further.

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [16]:
data['ctr'].describe()

,ctr
count,30000.000000
mean,0.510733
std,3.279162
min,0.000000
25%,0.000000
50%,0.070000
75%,0.290000
max,100.000000


In [18]:
#correlation between position and CTR
print(data[["avg_position", "ctr"]].corr())

              avg_position      ctr
avg_position       1.00000 -0.07259
ctr               -0.07259  1.00000


The raw correlation between `avg_position` and `ctr` is weak (r = -0.073)

In [19]:
#correlation between word count and impressions
print(data[["word_count", "impressions_90d"]].corr())

                 word_count  impressions_90d
word_count         1.000000         0.163345
impressions_90d    0.163345         1.000000


`word_count` shows only a weak positive correlation with `impressions_90d`
  (r = 0.163), suggesting content length alone is a modest signal at best,
  not a strong lever

In [20]:
#engagement by trend direction
print(data.groupby("trend_direction")["engagement_rate"].mean())

trend_direction
down      2.437193
flat      1.374514
new       2.135653
stable    2.838801
up        2.989578
Name: engagement_rate, dtype: float64


Engagement rate does not move in a simple pattern by `trend_direction`:
  "down" pages average 2.44 vs "flat" pages at 1.37. it counters to the
  assumption that declining pages always show worse engagement

In [21]:
#CTR by position tier
print(data.groupby("position_tier")["ctr"].mean().sort_values(ascending=False))

position_tier
top_3       1.483611
page_1      0.652467
striking    0.323239
page_3_5    0.222484
deep        0.150212
Name: ctr, dtype: float64


grouping CTR by `position_tier` reveals a much clearer pattern:
  mean CTR drops steadily from top_3 (1.48) to page_1 (0.65) to deep (0.15).
  This suggests the position-CTR relationship is not well captured by a
  linear correlation and may need a non-linear or grouped approach

## 4. Careful words: what I can and can't claim

**What I can say:**
- This work is *observational*, not experimental. I can describe
  associations between signals (like position, word count, freshness) and
  outcomes (like impressions, CTR, engagement) — not prove that one causes
  the other.
- I can say a signal is "associated with," "correlated with," or "linked
  to" an outcome, and describe how strong that relationship appears to be
  (e.g. weak, moderate) based on correlation coefficients or grouped
  comparisons.
- I can flag patterns that are counterintuitive or worth further
  investigation (e.g. declining pages showing higher engagement than flat
  pages) without claiming to explain why they happen.
- I can say this analysis is decision-support: it points a content
  strategist toward signals worth watching, not a guaranteed lever to
  pull.

**What I cannot say:**
- I cannot claim that changing a signal (e.g. increasing word count) will
  *cause* a page's performance to improve — that would require an
  experiment or causal design, which this data doesn't provide.
- I cannot claim to have reverse-engineered or predicted anything about
  Google's actual ranking algorithm. All I have is FlyRank's own observed
  search and engagement data, not access to how Google's systems work.
- I cannot treat a correlation as proof, especially a weak one (e.g. the
  -0.073 correlation between position and CTR) — a number that small could
  reflect noise, a non-linear relationship, or confounding factors I
  haven't controlled for.
- I cannot generalize findings from this ~30,000-row anonymized starter
  slice to the full ~79M-row warehouse without re-checking on that larger,
  more representative data.

## Self-check

Before you submit, confirm each line honestly:

- [Done] Every section above is filled — markdown thinking AND the code that backs it
- [Done ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [Done] No client names, URLs, or private queries anywhere
- [Done] My claims use careful words: observed, measured, directional, decision-support
- [Done] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.